stlab/forest_algorithms.hpp

```diff
@@ -0,0 +1,122 @@
+/**************************************************************************************************/
+
+#ifndef STLAB_FOREST_ALGORITHMS_HPP
+#define STLAB_FOREST_ALGORITHMS_HPP
+
+/**************************************************************************************************/
+
+// stdc++
+#include <optional>
+
+// stlab
+#include <stlab/forest.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+// "Congruent" would be a nice name here, but in geometry that also implies reflection.
+template <class Forest1, class Forest2>
+bool equal_shape(const Forest1& x, const Forest2& y) {
+    if (x.size_valid() && y.size_valid() && x.size() != y.size()) return false;
+    auto pos{y.begin()};
+    for (auto first(x.begin()), last(x.end()); first != last; ++first, ++pos) {
+        if (first.edge() != pos.edge()) return false;
+    }
+    return true;
+}
+
+/**************************************************************************************************/
+
+template <class Forest>
+struct forest_insert_iterator {
+    using value_type = typename Forest::value_type;
+    using iterator_type = typename Forest::iterator;
+
+    explicit forest_insert_iterator(Forest& f) : _f{f}, _p{_f.root()} {}
+
+    auto& operator*() { return *this; }
+
+    auto& operator++() {
+        ++_p;
+        return *this;
+    }
+
+    auto& operator=(value_type&& x) {
+        _p = _f.insert(_p, std::forward<value_type>(x));
+        return *this;
+    }
+
+    void trailing() { _p = trailing_of(_p); }
+
+private:
+    Forest& _f;
+    iterator_type _p;
+};
+
+template <class Forest>
+auto forest_inserter(Forest& f) {
```

Comment:
Is this necessary? In C++ 17 we have CTAD. If you want to have it so that it can be used before C++17, then I would prefer a name like make_forest_inserter. Then it would be consistent with make_pair, make_shared_ptr, etc.

Reply:
I've changed it to `forests::make_inserter`. I like that better, too.